### PR TITLE
Feat: Add PoissonNLL loss

### PR DESCRIPTION
**Pull request by salvomcl**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

This PR addresses issue #2751.

### Changes

This PR introduces the `PoissonNLLLoss` module, implementing the Poisson negative log-likelihood loss function. 
In particular I followed the`pytorch` implementation to make the transition from PyTorch as smooth as possible for users.

Key additions:  
- `PoissonNLLLossConfig`: Configuration structure with parameters for `log_input`, `full`, and `eps`.  
- `PoissonNLLLoss`: Implements forward and element-wise loss computation with optional reduction (`Mean`, `Sum`).  
- Validation checks to ensure appropriate input dimensions and non-negative target values.  

### Testing

- Added unit tests to verify:  
  - Correctness of the loss computation with different configurations.  
  - Behavior when `log_input` is `true` and `false`.  
  - Handling of the Stirling approximation term when `full` is enabled.  
  - Gradient computation correctness via backpropagation tests.  
  - Assertion failures for invalid inputs (negative targets, shape mismatches, and non-positive `eps`).

### CODE REVIEW

File: crates/burn-core/src/nn/loss/poisson.rs [hunk 1]
```diff
@@ -0,0 +1,389 @@
+use std::f32::consts::PI;
```
Comment (laggui) [inline crates/burn-core/src/nn/loss/poisson.rs:1]: For no-std we should use this instead

```suggestion
use core::f32::consts::PI;
```

File: crates/burn-core/src/nn/loss/poisson.rs [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+/// L(predictions, target) = predictions - target * log(predictions) + log(target!)
+/// ```
+/// The last term (`log(target!)`) can be omitted or approximated using Stirling's formula.
+/// The approximation is applied for `target > 1`, while for `target <= 1`, zeros are added to the loss.
+///
+/// For more details, see:
+/// <https://en.wikipedia.org/wiki/Poisson_regression#Maximum_likelihood-based_parameter_estimation>
+#[derive(Module, Debug, Clone)]
+#[module(custom_display)]
+pub struct PoissonNLLLoss {
```
Comment (laggui) [inline crates/burn-core/src/nn/loss/poisson.rs:87]: To follow the naming convention with other structs, this should be renamed to `PoissonNllLoss` (and the config to `PoissonNllLossConfig`, etc.)

File: crates/burn-core/src/nn/loss/poisson.rs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        loss_no_reduction.into_data().assert_approx_eq(&expected, 5);
+
+        let expected = TensorData::from([21.9920]);
+        loss.into_data().assert_approx_eq(&expected, 5);
+
+        let expected = TensorData::from([131.9518]);
+        loss_sum.into_data().assert_approx_eq(&expected, 5);
+    }
+
+    #[test]
```
Comment (laggui) [inline crates/burn-core/src/nn/loss/poisson.rs:289]: The `TestAutodiffBackend` type is under the `std` feature flag, so we probably need to guard this test.

```suggestion
    #[cfg(feature = "std")]
    #[test]
```

**Review by laggui — CHANGES_REQUESTED** [review]
LGTM! Thanks for the addition 🙂 

Just some minor changes left

**Comment by salvomcl** [comment]
Thanks for the review @laggui! Just pushed the requested changes! :)

**Review by laggui — APPROVED** [review]
